Migrate/unit staff editor 9.x

returnMarcco opened the pull request:
# Description

IMPORTANT: This PR is a replacement for this PR here: https://github.com/thoth-tech/doubtfire-web/pull/334/
The original migration was branched off of the `8.x` branch, and I had issues resolving merge conflicts. I was given permission to apply my changes to a branch created from the `9.x` branch. Please see the closed PR for Junior and Senior reviews of the migration, and comments from my mentor.

IMPORTANT: There are errors on the `9.x` branch when working with the `unit-staff-editor`. I can confirm that these errors occur in the `9.x` branch, and were not introduced by changes, as they still occur with the original Coffeescript implementation of the `unit-staff-editor`. Details will be left below.

This PR migrates the original implementation of the `unit-staff-selector` component directive from `Coffeescript`, `AngularJS` and `Angular Bootstrap` to `Typescript`, `Angular`, and `Angular Material`.

Fixes: 

https://teams.microsoft.com/l/entity/com.microsoft.teamspace.tab.planner/tt.c_19:bd20175d09414f079490a2403f7fca74@thread.tacv2_p_njykIFLDn0iAY1at7tACfcgADgBS_h_1714385623719?tenantId=d02378ec-1688-46d5-8540-1c28b5f470f6&webUrl=https%3A%2F%2Ftasks.teams.microsoft.com%2Fteamsui%2FpersonalApp%2Falltasklists&context=%7B%22subEntityId%22%3A%22%2Fv1%2Fplan%2FnjykIFLDn0iAY1at7tACfcgADgBS%2Ftask%2FB2wpDMIWEESKpHxSv3f4iMgAM6uL%22%2C%22channelId%22%3A%2219%3Abd20175d09414f079490a2403f7fca74%40thread.tacv2%22%7D

## Type of change

- [x] Migration
- [x] This change requires a documentation update (README.MD & Component Review)

# How Has This Been Tested?

### Follow the below steps to test the feature. Check the devtools console and network tab as you go. 
Note: There was an error that was present before this change. This pre-existing is stated in the `Notes` section below.
- Log into `onTrack` as a `Convenor`
- Click settings button (shield icon) at top right of the screen
- Click `Manage Units`
- Click the `Staff` menu-item
- Type in the searchbox. This should auto-complete and filter staff members by the provided search term
- Click a desired staff member to add them to the unit
- Ensure the added staff members do not appear in the staff list (filtering)
- Test removing a staff member by clicking the `bin` icon in the staff member row
- Test switching roles between `Convenor` and `Tutor
    - Add a person who is authorized to work as a Convenor (try `Reuben`)
    - Set them as a `Convenor` by clicking the `Convenor` button in the staff member row
    - Set them back to being a `Tutor`
 - Test assigning a `Main Convenor` for the unit by clicking the `Main Convenor` button in the staff member row
  
## Testing Checklist:

- [x] Tested in latest Chrome
- [ ] Tested in latest Safari (No access to a Mac for educational purposes)
- [x] Tested in latest Firefox

## Notes

- Existing errors on the `9.x` branch that impact the `unit-staff-editor` component
    - Units always seem to have `no staff` when loading the page. I can add and remove staff succesfully (correct HTTP response codes and data shows in the UI), however this does not persist between page loads. It seems to be a caching issue. This did not occur when I branched off the `8.x` branch. Please see this screenshot.
![Screenshot 2025-05-28 193046](https://github.com/user-attachments/assets/a2c57fc4-fe98-4c2c-96c1-b72d55d7e8d3)

    - An error thrown in a component that is higher up in the call stack. This occurs when changing the staff member's `role`. This error occurs even with the old implementation of the component when working on the `9.x` branch. It states an `undefined` `id` property.
 
- I have removed the `Add Staff` button that is present on the `8.0.x` branch as it seemed to do nothing except throw a type error. This was confirmed on the `8.0.x` branch. The only way I could find to add a staff member was through the autocomplete input field.
- I haven't used the `Angular Material` Table element as the migration guide states to scan for `bootstrap` components and swap them out with `Angular Material` components. The original implementation template seems to use a native HTML table.
- In the `8.0.x` branch, changing the `role` of a staff member threw a seemingly non-fatal type error. This error was present before my changes.

## Screenshots

Before Changes
![unit_staff_selector_before](https://github.com/user-attachments/assets/299a62a3-46f1-469b-a717-5383fa67c502)
![unit_staff_selector_before_2](https://github.com/user-attachments/assets/f262d55a-f4a9-4c12-969c-a0012936db21)

After Changes
![after_1](https://github.com/user-attachments/assets/b08b8549-7942-4347-942a-7ad91906ec6e)
![after_2](https://github.com/user-attachments/assets/c8c5e83a-7433-44cf-89c4-10195cfddf66)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have requested a review from @macite and @jakerenzella on the Pull Request

zara1504 committed:
chore: remove media-service.coffee and modify doubtfire-angularjs.module.ts

returnMarcco committed:
chore: migrate/unit-staff-editor

- delete old Coffeescript file
- delete old template
- remove reference to old Coffeescript file
- link new component
- downgrade new component

returnMarcco committed:
chore: migrate/unit-staff-editor

- port `unit-staff-editor-component` to Typescript from Coffeescript
- port `unit-staff-editor` template to Angular template syntax w/ Angular Material
- adjust parent component `Inputs` and attribute bindings

aNebula commented:
LGTM.
@returnMarcco Please proceed to open the upstream request against [doubtfire-lms/doubtfire-web 9.x branch](https://github.com/doubtfire-lms/doubtfire-web/tree/9.x)

returnMarcco committed:
Update README.md

returnMarcco committed:
Update README.md

returnMarcco committed:
Migrate/group set selector (#981)

* chore: migrate group-set-selector (WIP)

- define and declare new component files

- import new component into app

- downgrade component

* chore: migrate group-set-selector

- convert AngularJS template syntax to Angular syntax
- finish implementing `group-set-selector.component.ts`
- remove redundant Coffeescript file

* chore: migrate group-set-selector

- finish defining Typescript logic
- update parent component to use Angular syntax for child component declaration

* chore: cleanup and documentation

- remove redundant Coffeescript file
- adjust frontend migration todo list

* chore: remove redundant file

- remove redundant Coffeescript file

* fix: attribute and event binding

- fix attribute and event binding syntax
- remove redundant template file

* refactor: two-way binding

- remove redundant directive attributes

- rename EventEmitter var name in child component

* Update README.md

* Update README.md

* refactor: replace Bootstrap components with Angular Material

- replace Bootstrap components with Angular Material components & directives

- refactor to use modern Angular structural directives

returnMarcco committed:
refactor: migrate/date service (#931)

* chore: date-service.coffee to TS

- add `date.service.ts` file
- add `date.service.spec.ts` unit test
- downgrade TS module such that it works within AngularJS

* chore: unlink date-service.coffee

- complete unlinking process of  `date-service.coffee` after migrating to Typescript
- manual linting/comment formatting

* chore: delete redundant coffeescript file

- delete `date-service.coffee` file after migration

* chore:  formatting

- format code with prettier

lachlan-robinson committed:
refactor: migrate progress dashboard (#927)

* chore: 9.x windows

* chore: create new component files

* chore: unlink old component + link new component

* chore: migrate functionality + update to tailwindcss and material UI

* chore: remove old component files

* chore: responsive columns for visualisations

* Update package-lock.json

* Update package.json

* Update package.json

* Update package-lock.json

* chore: update styling to better match original

* chore: add custom breakpoint for responsive change

* chore: tweak breakpoint for better visualisation visibility

* chore: update tutor view functionality + tweak event emitter name

---------

Co-authored-by: Boink <40929320+b0ink@users.noreply.github.com>

lachlan-robinson committed:
fix: burndown chart visualisation (#942)

* fix: update burndown data to render series of different lengths

* chore: cleanup imports

* chore: remove leftover testing comment from previous contributor

b0ink committed:
Merge pull request #929 from zara1504/remove/media-service

Remove Media Service

lachlan-robinson committed:
refactor: migrate task status pie chart (#928)

* chore: migrate pie chart to ngxchart

* chore: remove testing console log

* chore: restore package and lock to 9.x

* chore: remove testing comments

* chore: resolve merge conflicts

* chore: remove testing console log

* chore: resolve futher conflicts

* chore: exlude packge-lock

* chore: exclude package

* chore: delete migrated progress dashboard html

---------

Co-authored-by: Boink <40929320+b0ink@users.noreply.github.com>

b0ink committed:
Merge branch '9.x' into migrate/unit-staff-editor-9.x
